### PR TITLE
Waf rules

### PR DESCRIPTION
**Pull request by oscarDovao**
This PR contains the implementation of the WAF rules related API endpoints. Rules cannot be created nor updated, only read, and the implementation provides a number of filters that can be used to customise the result set returned.

### CODE REVIEW

File: fastly/waf_rules.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
+package fastly
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/google/jsonapi"
+	"io"
+	"reflect"
+	"strconv"
+	"strings"
```
Comment (phamann) [inline fastly/waf_rules.go]: Please separate standard library imports from others.

File: fastly/waf_rules.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+			if value != 0 {
+				result[key] = strconv.Itoa(value.(int))
+			}
+		case "[]string":
+			if len(value.([]string)) > 0 {
+				result[key] = strings.Join(value.([]string), ",")
+			}
+		case "[]int":
+			if len(value.([]int)) > 0 {
+				result[key] = strings.Trim(strings.Join(strings.Fields(fmt.Sprint(value.([]int))), ","), "[]")
```
Comment (phamann) [inline fastly/waf_rules.go]: 🙈 can we make this more readable?

File: fastly/waf_rules.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+			}
+		}
+	}
+	return result
+}
+
+// ListWAFRules returns the list of VAF versions for a given WAF id.
```
Comment (phamann) [inline fastly/waf_rules.go]:
```suggestion
// ListWAFRules returns the list of VAF versions for a given WAF ID.
```

File: fastly/waf_rules.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	}, nil
+}
+
+// ListAllWAFRulesInput used as input for listing all WAF rules.
+type ListAllWAFRulesInput struct {
+	// Limit the returned rules to a set linked to a tag by name.
+	FilterTagNames []string
+	// Limit the returned rules to a set by publishers.
+	FilterPublishers []string
+	// Excludes individual rules by modsecurity rule ids.
```
Comment (phamann) [inline fastly/waf_rules.go]:
```suggestion
	// Excludes individual rules by modsecurity rule IDs.
```

File: fastly/waf_rules.go (hunk 1 above)
Comment (oscarDovao) [reply fastly/waf_rules.go]: imports organised

File: fastly/waf_rules.go (hunk 2 above)
Comment (oscarDovao) [reply fastly/waf_rules.go]: Looking at it now, I see what you mean :). I have refactored to make the implementation more explicit and clear

File: fastly/waf_rules.go (hunk 3 above)
Comment (oscarDovao) [reply fastly/waf_rules.go]: We use id id lowercase on similar comments

File: fastly/waf_rules.go (hunk 4 above)
Comment (oscarDovao) [reply fastly/waf_rules.go]: same as above

File: fastly/waf_rules.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+type ListWAFRulesInput struct {
+	// Limit the returned rules to a set linked to list of tags by name.
+	FilterTagNames []string
+	// Limit the returned rules to a set by publishers.
+	FilterPublishers []string
+	// Excludes individual rules by modsecurity rule IDs.
+	ExcludeMocSecIDs []int
+	// Limit the number of returned rules.
+	PageSize int
+	//Request a specific page of rules.
```
Comment (phamann) [inline fastly/waf_rules.go]:
```suggestion
	// Request a specific page of rules.
```

**Pull request merged**